### PR TITLE
Implement VerifiedBeacons in scraper

### PR DESCRIPTION
**Pull request by jamescowens**
This covers most of the bases and should leave you with only filling in anything I may have forgotten. I have done the serialization to put the VerifiedBeacons map as a "project" part in ScraperSendFileManifestContent, and I have implemented the deserialization in the ConvergedManifest from received manifests. I have also put in the plumbing to fill out the vector of verified beacons in the SB.

Note that the serialization/deserialization is a native serialization of the mVerifiedMap, and I felt it was unnecessary to use any of the heavyweight gzip compression/decompression. Note that I also felt it was unnecessary to persist the VerifiedBeacons map to a file.

### CODE REVIEW

File: src/scraper/scraper.cpp [hunk 1]
```diff
@@ -3828,47 +3919,88 @@ bool ScraperSendFileManifestContents(CBitcoinAddress& Address, CKey& Key)
...
 
-    CDataStream part(vchData, SER_NETWORK, 1);
+            ProjectEntry.last = 1;
 
-    manifest->addPartData(std::move(part));
+            manifest->projects.push_back(ProjectEntry);
+
+            CDataStream part(SER_NETWORK, 1);
+
+            WriteCompactSize(part, ScraperVerifiedBeacons.mVerifiedMap.size());
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: It doesn't look like we need to write the size header here, right? `CDataStream::operator<<` should call the serialization routine for `std::map` which writes the size internally.

File: src/scraper/scraper.cpp [hunk 2]
```diff
@@ -3828,47 +3919,88 @@ bool ScraperSendFileManifestContents(CBitcoinAddress& Address, CKey& Key)
...
+
+        filein.fclose();
+
+        // The first part number will be the BeaconList.
+        manifest->BeaconList = iPartNum;
+        manifest->BeaconList_c = 0;
+
+        CDataStream part(vchData, SER_NETWORK, 1);
+
+        manifest->addPartData(std::move(part));
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: I think we need to `std::move(vchData)` in the `CDataStream` constructor for the next one to have an effect.

File: src/scraper/scraper.cpp [hunk 3]
```diff
@@ -4699,6 +4852,30 @@ bool LoadBeaconListFromConvergedManifest(const ConvergedManifest& StructConverge
...
+    {
+        CDataStream part(iter->second, SER_NETWORK, 1);
+
+        part >> VerifiedBeaconMap;
+
+        for (const auto& entry : VerifiedBeaconMap)
+        {
+            CPubKey PubKey;
+
+            CKeyID KeyID = PubKey.Parse(entry.first).GetID();
```
Comment (cyrossignol) [inline src/scraper/scraper.cpp]: The map is keyed by `CKeyID` (`uint160`). I need to go back and check... I think we can just use `KeyID.SetHex(entry.first)` here.

**Comment by cyrossignol** [comment]
I agree&mdash;I don't think we need to gzip this stuff.

File: src/scraper/scraper.cpp (hunk 1 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: i think you are right on that.

File: src/scraper/scraper.cpp (hunk 3 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: I will go back and check, but I think we stringified the scraper side maps. That is one reason I renamed them. I was getting confused between the scraper versions and the ones in the beacon class.

**Comment by jamescowens** [comment]
I have been thinking about converting the data structures over to native now that you have fixed the rest of the wallet. It is a pretty big job, and we have to make some decisions about the on disk format of the scraper files. I specifically used csv.gz so that you could easily decompress and load them into your spreadsheet of choice directly. We could move towards native structures and decompress and convert from csv to native when loading the files to send, and the reverse when storing. Most of the compression that gzip is doing is simply compensating for the inefficiency in the string based storage I think. BTW, this would have the effect of largely confining the gzip/unzip load to the scrapers themselves.

File: src/scraper/scraper.cpp (hunk 2 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: Hmm. I haven't done that in other places, but ok.

File: src/scraper/scraper.cpp (hunk 3 above)
Comment (jamescowens) [reply src/scraper/scraper.cpp]: typedef std::map<std::string, ScraperPendingBeaconEntry> ScraperPendingBeaconMap;

and from BeaconConsensus GetConsensusBeaconList()...

    for (const auto& pending_beacon_pair : pending_beacons)
    {
        const CKeyID& key_id = pending_beacon_pair.first;
        const NN::PendingBeacon& pending_beacon = pending_beacon_pair.second;

        if (pending_beacon.m_timestamp >= max_time)
        {
            continue;
        }

        ScraperPendingBeaconEntry beaconentry;

        beaconentry.timestamp = pending_beacon.m_timestamp;
        beaconentry.cpid = pending_beacon.m_cpid.ToString();

        consensus.mPendingMap.emplace(key_id.ToString(), std::move(beaconentry));
    }

So on the scraper side, the pending beacon map is keyed by the string of the CPID.

Comment (jamescowens) [reply src/scraper/scraper.cpp]: Or are you talking about the superblock/beacon side map here. Maybe I am confused...
Yep. You're right. I am confused. You can go directly to the uint160 representation by doing SetHex on the string representation. Duh.

**Review by cyrossignol — APPROVED** [review]

**Pull request merged**